### ISSUE TITLE
Houdini or Maya export

### ISSUE THREAD

**Issue by sabotage3d** [reattiva/Urho3D-Blender#39]
Hello I am thinking of using your code for making a Houdini or Maya export tool for Urho3d . Is there a simple way to seperate the Blender specific code from the Urho3d specific code. 
I am familiar with Maya and Houdini API's in python I just need a bit of direction on which classes exactly do I need. I think the trickiest would be the weights and bones animations. 
If you can provide a command line example on how to feed some data that would be great.

Thanks in advance for any help.

Alex

**Comment by sabotage3d** [comment]
Hey man do you have a simple example on how to use UrhoWriteModel(model, filename)
A simple triangle example would be great.

**Comment by reattiva** [comment]
```
def WriteTriangleExample():
    uModel = UrhoModel()
    uModel.name = "triangle"

    uGeometry = UrhoGeometry()
    uModel.geometries.append(uGeometry)

    uLodLevel = UrhoLodLevel()
    uLodLevel.distance = 0.0
    uLodLevel.primitiveType = TRIANGLE_LIST
    uLodLevel.vertexBuffer = 0
    uLodLevel.indexBuffer = 0
    uLodLevel.startIndex = 0
    uLodLevel.countIndex = 3
    uGeometry.lodLevels.append(uLodLevel)

    vertexBuffer = UrhoVertexBuffer()
    vertexBuffer.morphMinIndex = 0
    vertexBuffer.morphMaxIndex = 0
    uModel.vertexBuffers.append(vertexBuffer)

    indexBuffer = UrhoIndexBuffer()
    indexBuffer.indexSize = 2
    uModel.indexBuffers.append(indexBuffer)

    tVertex = TVertex()
    tVertex.pos = Vector((0.0, 0.0, 0.0))
    tVertex.normal = Vector((0.0, 0.0, -1.0))

    uVertex = UrhoVertex(tVertex)
    vertexBuffer.updateMask(uVertex.mask)
    vertexBuffer.vertices.append(uVertex)
    indexBuffer.indexes.append(0)
    uModel.boundingBox.merge(uVertex.pos)
    uGeometry.center += uVertex.pos;

    tVertex = TVertex()
    tVertex.pos = Vector((0.0, 1.0, 0.0))
    tVertex.normal = Vector((0.0, 0.0, -1.0))

    uVertex = UrhoVertex(tVertex)
    vertexBuffer.updateMask(uVertex.mask)
    vertexBuffer.vertices.append(uVertex)
    indexBuffer.indexes.append(1)
    uModel.boundingBox.merge(uVertex.pos)
    uGeometry.center += uVertex.pos;

    tVertex = TVertex()
    tVertex.pos = Vector((1.0, 0.0, 0.0))
    tVertex.normal = Vector((0.0, 0.0, -1.0))

    uVertex = UrhoVertex(tVertex)
    vertexBuffer.updateMask(uVertex.mask)
    vertexBuffer.vertices.append(uVertex)
    indexBuffer.indexes.append(2)
    uModel.boundingBox.merge(uVertex.pos)
    uGeometry.center += uVertex.pos;

    uGeometry.center /= 3;

    UrhoWriteModel(uModel, "e:/temp/triangle.mdl")
```

**Comment by sabotage3d** [comment]
Works like a charm. Thanks a lot :+1:

**Comment by sabotage3d** [comment]
I am getting this error when I run the example. I just bypassed the ```  raise VertexMaskError(oldMask, vertexMask) ``` . Would that cause any issues ?

```
raise VertexMaskError(oldMask, vertexMask)
export_urho.VertexMaskError: 0002 vs 0003
 differences: Position
```

**Comment by reattiva** [comment]
The vertex buffer only accepts vertices with the same attributes.
You have added to the vertex buffer a vertex with only the normal attribute (0x0002) then you try to add a vertex with normal and position attributes (0x0002 | 0x0001).

**Comment by sabotage3d** [comment]
I think I am making a mistake somewhere.
Let say you have 16 vertices, 8 positions, 16 indices, 16 vertex normals .
Normally we would send 8 positions and 16 indices to represent 16 vertices but we have 16 unique normals do we have to pass then 16 positions, 16 indices and 16 vertex normals ?

**Comment by reattiva** [comment]
That's right, a different normal means a different vertex even if its position is the same.

**Comment by sabotage3d** [comment]
Isn't that inneficient ? Having duplicate positions of vertices that share the same index ?

**Comment by reattiva** [comment]
Over position and normals, you also have UV, tangents, colors ... you can add anything you want to a vertex. I don't know if it is efficient, as you are suggesting, defining a vertex as a set of indices one for each of these attributes.
Make some research but as far as I know, with a VBO you reuse a vertex only if all its attributes are the same.

**Comment by sabotage3d** [comment]
thanks man. It all makes sense now.

**Comment by sabotage3d** [comment]
I am trying to understand the skeleton data now. I can't fully understand how to do the skinning properly. Do you have a simple example on how to pass a single bone to the MDL ?
What exactly is the ``` float[12]  4x3 offset matrix``` suppose to do  ? Where do we pass vertices weights to specific bone ?

**Comment by reattiva** [comment]
Unfortunately I don't have an example for that.
You pass weights and indices as attributes of a vertex. The index points to the bone array of the model, the weight indicate how strong the vertex position will be affected. Each vertex  can be influenced by 4 bones at maximum. The total number of bones affecting a subgeometry should be less that 64, there is a remapping system for having a skeleton with more that 64 bones and then each subgeometry is remapping for having (when possible) less than that. However I think the 64 bones limit was changed lately.
The offset matrix should be the bone transformation (rotation + position + scale) in object space. Normally in the skeleton you specify the position and rotation of a bone in respect of its parent, this matrix specifies the same data but in object space (in respect of its origin).
And yes, in the 'ani' there is a track for each bone and the association happens by name.
I went by memory so take all of this with a grain of salt.

**Comment by sabotage3d** [comment]
Thanks a lot for all the help I managed to get most of the things working. I am having problems with flipping bones. Is there a specific stable method to convert from world space matrix of a bone to the offset matrix for Urho3d. So far I was multiplying by the inverse matrix of a parent bone then doing quaternion and negating the frist two components but I am getting flips.

**Comment by reattiva** [comment]
I must be honest, I don't remember how it works anymore. However re-reading my comments I think the "offset matrix" should be relative to the armature and not to the parent bone, in other words it should be in object space, so you shouldn't be multiplying by the inverse.

**Comment by sabotage3d** [comment]
In Houdini I am getting world space transformation of a bone to bring it to object space I am multiplying it by inverse of the parent of this bone, which brings it back to centre in object space. Let me know if I am doing something wrong ?

**Comment by reattiva** [comment]
If you multiply the world space transformation (rotation+translation) matrix of a bone B with the inverse of world space matrix of its parent bone P, you get the transformation of B relative to P. This is not the object space. If you armature A is at the center of the world and it's not rotated, the world space matrix of the bone B is the object space matrix, instead if A is traslated or rotated, then you multiply the world space matrix of B with the inverse of the world space matrix of A, and you get the trasformation of B relative to A, that is the object space matrix of B.

**Comment by sabotage3d** [comment]
So the skinning offset matrix should be relative to the root of the skeleton. What about the actual quaternion and translation of a bone? In your code it looks like they are relative to its parent ?
```
        if parent:
            boneMatrix = parent.matrix_local.inverted() * boneMatrix
        else:
            boneMatrix = originMatrix * boneMatrix
```

**Comment by reattiva** [comment]
'boneMatrix' is not the skinning offset matrix, try to follow the code more carefully, start with 'ml' (just below your quoted code), then 'worldTransform' in the TBone class, then 'inverseMatrix' in UrhoBone class, this is the matrix written to file. Again, trust me not.

**Comment by sabotage3d** [comment]
Yeah I know that it is not the offset skinning matrix. I already got the offset matrix working the whole flipping seems to be caused by the quaternion which I think I am doing exactly like in your code. If we have bone1 and bone2 and bone1 is the parent of bone 2. Then I am doing bone2.world_matrix * bone1.world_matrix.inverted() then from this matrix I am doing the quaternion, the root is at 0,0,0. Is there something I am missing ?

**Comment by reattiva** [comment]
That is bone2 relative to bone1.
Anyway, try to print the matrices in my last post and see where they don't match with yours.

**Comment by sabotage3d** [comment]
Thanks you are right apprently in Houdini every bone can have different orientation which is causing the issues. It will take me a while to fix it.
